spec.md

```diff
@@ -1586,12 +1760,192 @@ en <= Z_valid
 cover(clk, pred, en, "X equals Y when Z is valid") : optional_name
 ```
 
+## Probes
+
+Probe references are created with `probe`{.firrtl} expressions, exported with
+the `export`{.firrtl} statement, forwarded between instances using the
+`forward`{.firrtl} statement, read using the `read`{.firrtl} expression (see
+[@sec:reading-probe-references]), and used with `force`{.firrtl} and
+`release`{.firrtl} statements.
+
+Export and forward are used to route references through the design, and may be
+used wherever is most convenient in terms of available identifiers.
+Every sink-flow probe must be the target of exactly one of these statements.
+
+These statements are detailed below.
+
+### Export
+
+The export statement takes a reference expression, which must be a probe expression,
+and exports the probe reference to into a sink-flow static reference target.
+
+```firrtl
+module Refs:
+  input clock:  Clock
+  output a : Probe<{x: UInt, y: UInt}> ; read-only ref. to wire 'p'
+  output b : RWProbe<UInt> ; force-able ref. to node 'q', inferred width.
+  output c : Probe<UInt<1>> ; read-only ref. to register 'r'
+  output d : RWProbe<UInt<2>>[4] ; vector of ref.'s to memory data in 'm'
+  output e : Probe<Clock> ; ref. to input clock port
+
+  wire p : {x: UInt, flip y : UInt}
+  export probe(p) as a ; probe is passive
+  node q = UInt<1>(0)
+  export rwprobe(q) as b
+  reg r: UInt, clock
+  export probe(r) as c
+  mem m:
+    data-type => UInt<5>
+    depth => 4
+    ; ...
+    read-under-write => undefined
+
+  export rwprobe(m) as d
+  export probe(clock) as e
+```
+
+Exporting to a field within a bundle or other statically known sub-element of
+an aggregate is allowed, for example:
+```firrtl
+module Foo:
+  input x : UInt
+  output y : {x: UInt, p: Probe<UInt>}
+  output z : Probe<UInt>[2]
+
+  wire p : UInt
+  p <= x
+  y.x <= p
+
+  export probe(p) as y.p
+  export probe(p) as z[0]
+  export probe(p) as z[1]
+```
+
+`RWProbe`{.firrtl} references to ports are not allowed on public-facing modules.
+
+#### Probes and Passive Types
+
+While `Probe`{.firrtl} inner types are passive, the type of the exported expression
+is not required to be:
+
+```firrtl
+module Foo :
+  input x : {a: UInt, flip b: UInt}
+  output y : {a: UInt, flip b: UInt}
+  output xp : Probe<{a: UInt, b: UInt}> ; passive
```

Comment:
Yeah let's sync up in a meeting and report back.

Reply:
Summary of suggested alternative after discussion:

Push connect-all-regardless-of-alignment to `read`/`force`, Probe's have whatever type+alignment the originating target does.  This pushes that behavior to `read`/`force` where in the current proposal this is done by the `probe`/`rwprobe` expression creating the `Probe`.

As I understand it, in nearly all cases this is semantically equivalent to the current proposal (only difference is that syntactically the Probe types have whatever alignment/full type the target does).
So it comes down to preference re:aesthetics/perspective I think and perhaps interaction with future ideas/directions (?).  Without advocating one way or another, as an observation: in a future where we might have more powerful ways to project/view/connect bundles the difference may be more significant/relevant.

One phrasing of the difference here:
is it nicer for inner type to always match the probe target (even if that doesn't help/change anything), or should the inner type reflect the type of the probed data you interact with (always passive, not necessarily the same as type of the target)?  If/when we support references to circuit elements directly (not probes of their data) I think those types should definitely match the originating type (something like: `Wire<{a : UInt, flip b: UInt}>` or more usefully a Memory, Instance, or Module).

One result of keeping the alignment would be that now the probes for two entities that have identical types other than alignment are no longer equivalent, even though they describe the same data and are interacted with in the same way (FWIW).  While we could say they're compatible, I think if that's the case we shouldn't bother recording the flip-ness anyway.  I don't think this is especially important, but would be one difference.

Reply:
Excellent summary. I think the main advantage to me is that you are preserving as much type information as possible, for as long as possible. Since reference types will be used at the interface level, I think including that information seems useful for future purposes. For example, I could imagine a "force only the flipped fields" as an API a user may want to support in the future in Chisel, which would require preserving the relative field alignments. E.g. using effectively Chisel's `:<=` operator, but with forces. Semantically, this would be to "force only the data but not the backpressure fields".